File: exercise/ts/day11/src/christmas/preparation.ts
import {ToyType} from "./toyType";

export class Preparation {
    static prepareGifts(numberOfGifts: number): string {
        if (numberOfGifts <= 0) {
            return 'No gifts to prepare.';
        } else if (numberOfGifts < 50) {
            return 'Elves will prepare the gifts.';
        }
        return 'Santa will prepare the gifts.';
    }

    static categorizeGift(age: number): string {
        if (age <= 2) {
            return 'Baby';
        } else if (age <= 5) {
            return 'Toddler';
        } else if (age <= 12) {
            return 'Child';
        }
        return 'Teen';
    }

    static ensureToyBalance(toyType: ToyType, toysCount: number, totalToys: number): boolean {
        const typePercentage = toysCount / totalToys;

        switch (toyType) {
            case ToyType.EDUCATIONAL:
                return typePercentage >= 0.25;
            case ToyType.FUN:
                return typePercentage >= 0.30;
            case ToyType.CREATIVE:
                return typePercentage >= 0.20;
            default:
                return false;
        }
    }
}

File: solution/day11/Christmas/Preparation.cs
namespace Christmas
{
    public enum ToyType
    {
        Educational,
        Fun,
        Creative
    }

    public static class Preparation
    {
        public static string PrepareGifts(int numberOfGifts)
            => numberOfGifts switch
            {
                <= 0 => "No gifts to prepare.",
                < 50 => "Elves will prepare the gifts.",
                _ => "Santa will prepare the gifts."
            };

        public static string CategorizeGift(int age)
            => age switch
            {
                <= 2 => "Baby",
                <= 5 => "Toddler",
                <= 12 => "Child",
                _ => "Teen"
            };

        public static bool EnsureToyBalance(ToyType toyType, int toysCount, int totalToys)
            => ((double) toysCount / totalToys)
                .Do(typePercentage =>
                    toyType switch
                    {
                        ToyType.Educational => typePercentage >= 0.25,
                        ToyType.Fun => typePercentage >= 0.30,
                        ToyType.Creative => typePercentage >= 0.20,
                        _ => false
                    });
    }
}